[WIP] "abstract_ui" fixes and improvements

tarruda opened the pull request:
FWIW I noticed these things while prototyping a UI in python

tarruda committed:
test: Fix eol_clear in screen.lua

It should only clear to the end of the current scroll region(this is the
behavior expected by nvim)

tarruda committed:
ui: Add update_fg/update_bg methods

It is necessary to notify the UI when the default background/foreground colors
change in order to render correctly.

tarruda committed:
ui: Increase cursor row when text being rendered would cross its limit

tarruda committed:
ui: Add mising T_CSV code to abstract_ui termcap

This code is required for screen.c to set a vertical scroll region, which is a
great optimization when scrolling vertically split windows.

tarruda committed:
ui: Test for abstract_ui whenever a minimal t_colors value is required

t_colors should not be checked when abstract_ui is active, because nvim UI is
not limited to a terminal.

tarruda committed:
remote_ui: Move handler registration to remote_ui.c

Also don't defer attach_ui handling

tarruda committed:
ui: Fix ui resizing and change some method names

tarruda committed:
test: Remove unnecessary command from highlight_spec

tarruda committed:
input: Ignore invalid "<" key sequences

Ignoring invalid key sequences simplifies input handling in UIs. The only
downside is having to use "<lt>" everytime a "<" is needed on functional tests.

tarruda committed:
ui: Merge standout and reverse into one attribute

tarruda commented:
I have pushed a gtk+ GUI prototype to ui-prototypes branch in python client. It is necessary to install nvim from this branch and python gobject introspection bindings, on a debian/ubuntu system this should take care of installing the python stuff:

```
sudo apt-get install python-gi
sudo pip install git+https://github.com/tarruda/python-client.git@ui-prototypes
```

The GUI is started with the `pynvim` command.

The prototype is almost complete(mouse, keys, resizing..), any visual problems are probably due to nvim bugs that will be fixed here(bug reports are welcome in this PR)

tarruda committed:
syntax: Take rgb fg/bg when allocating cterm attr number

tarruda committed:
runtime: Fix plugin/matchparen.vim for abstract_ui

carlosdcastillo commented:
I installed the correct version of neovim:

(neovim)carlos@iapetus:~/vc/neovim$ /Users/carlos/usr/new-dec-2014-abstract-ui-fixes/bin/nvim --version
NVIM 0.0.0-alpha+201501022116 (compiled Jan  3 2015 02:40:03)
Commit: b430078047256810fed734661b8b2bf2e4c32977

I installed the correct version the python client, using: pip install git+https://github.com/tarruda/python-client.git@ui-prototypes

I installed gtk3 and py27-gobject3 in macports. I then run:

(neovim)carlos@iapetus:~$ pynvim --nvim-command=/Users/carlos/usr/new-dec-2014-abstract-ui-fixes/bin/nvim

and it just sits there. Can you please help me determine what I'm doing wrong?

tarruda commented:
@carlosdcastillo is the alternate nvim installation being resolved on PATH correctly?(you can confirm with `which nvim`).

Another option is to pass `--nvim-command='$FULL_NVIM_PATH --embed'` to the `pynvim` command

justinmk commented:
In terminal (haven't tried GUI), adding `set lines=999 columns=999` to `.nvimrc` causes broken display, even with this PR.

justinmk commented:
`:intro` doesn't redraw the screen, so if you have multiple windows open, the statuslines etc. appear behind the intro text. It's kind of cool, but technically a regression.

tarruda commented:
> In terminal (haven't tried GUI), adding set lines=999 columns=999 to .nvimrc causes broken display, even with this PR.

This also seems to happen with vanilla vim 7.4.316 (terminal UI)

> :intro doesn't redraw the screen, so if you have multiple windows open, the statuslines etc. appear behind the intro text. It's kind of cool, but technically a regression.

I couldn't reproduce this after a few quick tries, what were the exact steps you followed?

justinmk commented on runtime/plugin/matchparen.vim:
```diff
@@ -45,7 +45,7 @@ function! s:Highlight_Matching_Pair()
   " Avoid that we remove the popup menu.
   " Return when there are no colors (looks like the cursor jumps).
-  if pumvisible() || (&t_Co < 8 && !has("gui_running"))
+  if pumvisible() || (&term != 'abstract_ui' && &t_Co < 8 && !has("gui_running"))
```
Is this a temporary workaround? Shouldn't `t_Co` return the correct number for the currently active TUI?

tarruda replied:
Yes and no. While abstract_ui is temporary(it will be removed once the new TUI migration is complete), I'm still not sure how to deal with the `&t_Co` option in vimscript.

Most of the times, this is checked to decide what to display based on the number of terminal colors, but since nvim can have multiple terminals attached, this simply doesn't make sense. 

In the C code where t_Co is checked, I will simply remove the conditions and hardcode the path where most colors are available, but for vimscript I see two choices:

- Add a stub value of 256 to `&t_Co`
- Require that existing vimscript testing for `&t_Co` to be modified to take nvim into consideration.

Both approaches will probably require color schemes designed for multiple UIs to be adapted for nvim to work correctly. For example, in a color scheme that supports terminals with 16, 88 and 256 colors the current solution is something like this:

```vim
if &t_Co == 256
  highlight GROUP ctermfg=x
else if &t_Co == 88
  highlight GROUP ctermfg=y
else
  highlight GROUP ctermfg=z
endif
```

while for nvim the ideal approach would be:

```vim
highlight GROUP term256fg=x term88fg=y term16fg=z
```

Which would display correctly if multiple TUIs with different color count are connected at the same time(each TUI would receive the complete highlight information and use what is appropriate)

Clearly this would require [this change](https://github.com/neovim/neovim/issues/1767#issuecomment-68656372) to work.

justinmk commented on src/nvim/ui.c:
```diff
@@ -323,11 +341,10 @@ static void set_highlight_args(int mask, HlAttrs *attrs)
   attrs->undercurl = mask & HL_UNDERCURL;
   attrs->italic = mask & HL_ITALIC;
-  attrs->reverse = mask & HL_INVERSE;
+  attrs->reverse = mask & (HL_INVERSE | HL_STANDOUT);
```
curious about why this was necessary

tarruda replied:
Standout is almost always the same as reverse, but it is a very terminal specific term. Also, I don't think it would make sense to ever pass both standout and reverse to a terminal, so I thought it would be better to have a single attribute and let TUI decide to use standout when reverse mode is not available.

That said, I think we should go in [this direction](https://github.com/neovim/neovim/issues/1767#issuecomment-68656372) and allow arbitrary key/value pairs in the highlight command, which would move knowledge about text attributes from nvim core to UIs

tarruda committed:
screen: Remove abstract_ui test in reset_cterm_colors

tarruda committed:
screen: Fix bug that causes separator to always be redraw for utf8

carlosdcastillo commented:
I've been testing this patch and it is looking really good. I have some comments and some possible issues I've found:

* If you use vim_feedkeys and q, r or f it breaks the communication with the ui. As if you did ui_detach or something. This doesn't happen with vim_input (which is what you use in pynvim). What is the vim_feedkeys use case?

* I don't understand set_scroll_region(a,b,c,d) does. When scrolling in one direction it gets one set of parameters as if to say "prepare these currently out of view lines for scrolling" but in the other direction I think it does not happen. 

* Your client (pynvim) feels very fast. There are a variety of things you do: the thread/greenlet to read, the red black trees to track the dirty pixels to redraw, changed to gtk. Which of these are the ones you feel you got the most benefit from?

* In your client when I press k to go up it doesn't redraw (the first line  gets updated and the next line redrawn all over the screen).

bfredl commented:
I had similar experince with `feedkeys` (when experimenting with REPL integration). Intrestingly `cw` works, but `ciw` do not (feeding one byte at a time). I used `feedkeys` because somehow I thought it was better for feeding raw ASCII terminal input bytes, but it turns out `vim_input` also handles raw CTRL bytes and unescaped `<`. Switching to `vim_input` it all works much smoother. Perhaps the api documentation could explain better what "`[vim_input] will use a lower-level input buffer and the call is not deferred.`" and "`the string needs escaping for K_SPECIAL/CSI bytes`" actually implies for the designer of a ui front end?

tarruda commented:
> If you use vim_feedkeys and q, r or f it breaks the communication with the ui. As if you did ui_detach or something. This doesn't happen with vim_input (which is what you use in pynvim). What is the vim_feedkeys use case?

`vim_feedkeys` is the API version of the `feedkeys()` vimscript function. It is there more for compatibility, I intend to remove it later and provide alternative ways of controlling mapping behavior. In any case, it should not cause the UI to disconnect, I will investigate.

> I don't understand set_scroll_region(a,b,c,d) does. When scrolling in one direction it gets one set of parameters as if to say "prepare these currently out of view lines for scrolling" but in the other direction I think it does not happen.

set_scroll_region is called before scrolling a region of the screen in any direction, it is used a lot when scrolling split windows. I added this function because screen.c currently uses it in the form of termcap codes. Once I remove the current terminal UI code there will be improvements to the UI protocol(stateful functions such as cursor_goto, highlight_set, set_scroll_region will be removed)

> Your client (pynvim) feels very fast. There are a variety of things you do: the thread/greenlet to read, the red black trees to track the dirty pixels to redraw, changed to gtk. Which of these are the ones you feel you got the most benefit from?

You must update it, I no longer use the btrees package. The only reason I spawn another thread is to run the gtk event loop in parallel with libuv event loop(which handles nvim communication), speed has little to do with it. In fact, python GIL makes the use of threading pointless, I think it would be little faster if I used a single thread with one event loop controlling the entire application. Also, I no longer track dirty regions. I may have been doing something wrong, but the UI was more responsive when I simply redraw everything.

Since python is a slow language, I had one goal when building that UI: Minimize the amount of python code executed, and the gtk/cairo API lets me do that(the bulk of the work is done gtk+ C code). The most expensive part of the python code is scrolling the internal screen. If you use a very small font, there will be a noticeable slowdown(scrolling is O(n * m)). If you are working in a high level language, try to use the native graphical API to perform as much work as possible(also minimize the number of calls).

> Perhaps the api documentation could explain better what "[vim_input] will use a lower-level input buffer and the call is not deferred." and "the string needs escaping for K_SPECIAL/CSI bytes" actually implies for the designer of a ui front end?

It boils down to this: Deferred API calls result in `K_EVENT`(special key designed to integrate asynchronicity) being returned to nvim and are only handled in event loops that surround the `vgetc()` call with `event_enable_deferred()`/`event_disable_deferred()`. Since special keys are returned in every deferred event, multiple `vim_feedkeys` calls will be interleaved with `K_EVENT`, which can break pending operators.

tarruda committed:
ui: Dont resize screen if no UIs are attached

This prevents a race condition when a UI attaches early in the program and can
receive redraw commands for a invalid screen

tarruda committed:
ui: Add 'rgb' parameter to ui_attach

When set to false, nvim will send cterm color numbers with `highlight_set`.

oni-link commented:
Is installing the python-client in a specific directory supported? 
```
pip install -t . git+https://github.com/tarruda/python-client.git@ui-prototypes
```
Cannot find the file pynvim.

tarruda commented:
> pip install -t . git+https://github.com/tarruda/python-client.git@ui-prototypes

It seems the pynvim command-line wrapper isn't installed with the `-t` flag, but you can invoke the entry point directly with `python -m neovim.ui.cli`(note that -g is required for the gtk interface)

If you don't want to pollute system packages I recommend [pyenv](https://github.com/yyuu/pyenv), [here's](https://gist.github.com/tarruda/b67472898a7e4bd81572) a script that automates pygtk3 installation into a pyenv(you still need gtk3 installed system wide)

tarruda committed:
channel: Free channels created from libuv streams

tarruda committed:
ui: Don't parse abstract_ui codes if there are no attached UIs

tarruda committed:
ui: Reimplement `:suspend` command for remote UIs.

- Remove suspend method from the UI protocol
- Handle `:suspend` by disconnecting the last channel that sent a request to
  nvim.

tarruda commented:
Almost done with this, I will write some tests and then merge this with neovim/python-client#74

Some highlights on the latest versions:

- Now there are gtk and tickit prototypes. By default `pynvim` will start a terminal UI, but gtk can be chosen with the `--gui/-g` flag
- The `--listen/-l` and `--connect/-c` options have been added. They will spawn nvim independent from the parent python process, so it is possible to detach/attach to headless nvim instances. Not passing `--listen` or `--connect` will result in a embedded nvim instance being spawned.
- `:suspend` semantics have been adapted for remote uis: It will disconnect the UI that sent the mapping or command that resulted in `:suspend` being executed. This may change later but for now it is the simplest workaround I could find

tarruda commented:
BTW, those having trouble installing libtickit(the upstream bzr version doesn't seem to be building cleanly) can use the version bundled with nvim by setting `PKG_CONFIG_PATH` the first time `pynvim` is executed. For example, if nvim source code with deps is located at `$HOME/nvim`, this should make everything work(assuming cwd is the python-client source directory):

```bash
rm -rf neovim/ui/__pycache__
PKG_CONFIG_PATH=$HOME/nvim/.deps/usr/lib/pkgconfig pynvim  # <- will compile/link a native python module the first time it is executed
```

I'd appreciate if other neovim users helped me test pynvim, especially the tickit UI which will soon be converted to C and ported to neovim(in another PR). To those interested in helping, consider adding these aliases to your shell rc:

```bash
alias gvim='pynvim -g --'
alias vim='pynvim --'
```

fwalch commented:
The (terminal) GUI terminates immediately most of the time when I hit e.g. h, j, k, or l in an empty buffer. For a split second, I can see some traceback (looks like the error occurs in Python), but I can't read it that quickly.
Going into insert mode and inserting some text works,. That doesn't happen with `-u NONE`, but I couldn't figure what causes this exactly. Is there some way to see a debug/crash log?

tarruda commented:
> Is there some way to see a debug/crash log?

Try commenting the line that contains `TICKIT_TERMCTL_ALTSCREEN` so it will use the same screen buffer

fwalch commented:
Thanks @tarruda!

The output is:

```
    Traceback (most recent call last):
  File "/home/florian/.local/lib/python2.7/site-packages/neovim/ui/ui_bridge.py", line 78, in apply_updates                                                    
    handler = getattr(self._ui, '_nvim_' + update[0])                                                                                                          
AttributeError: 'TickitUI' object has no attribute '_nvim_visual_bell'                                                                                         
```

tarruda commented:
@fwalch it should be fixed now(though screen flashing isnt implemented in the gtk UI yet)

bfredl commented:
Just adding `PKG_CONFIG_PATH` didn't do the trick for me, but I managed to both build and run the terminal ui using
`LD_LIBRARY_PATH=$HOME/dev/neovim/.deps/usr/lib LIBRARY_PATH=$HOME/dev/neovim/.deps/usr/lib PATH=~/dev/neovim/build/bin:$PATH PKG_CONFIG_PATH=$HOME/dev/neovim/.deps/usr/lib/pkgconfig python2 -m neovim.ui.cli`
It doesn't work with my "csapprox" color scheme  (but this is partly my own fault, as my color scheme lanches a python script that manipulates the color palette of urxvt, which somehow causes the frontend to crash) but intrestingly enough, scolling is actually faster than in the built in terminal ui :+1:

bfredl commented:
Also, when exiting, it doesn't restore terminal state, i e mouse events continue to get sent.

fwalch commented:
@tarruda Can confirm that visualbell doesn't cause problems anymore.
I'll keep using this terminal GUI instead of `nvim`. Keep up the great work! :+1:

fwalch commented:
The `<` character doesn't get sent to the UI (cannot insert it, cannot use it in commands [indentation]). `>` does work. I'm using a German keyboard layout.

tarruda commented:
> Also, when exiting, it doesn't restore terminal state, i e mouse events continue to get sent.

Added a Force/disable mouse when exiting. @Leonerd, is this something libtickit should do automatically when `tickit_term_destroy` is called?

> The < character doesn't get sent to the UI somehow (cannot insert it in text, cannot use it in commands [indentation], ..). > does work. I'm using a German keyboard layout.

Should be fixed now

leonerd commented:
@tarruda: Yeah, Tickit should tidy up the terminal again when you shut it down. If it doesn't, that's a bug. I'll take a look

fwalch commented:
> Should be fixed now

:+1:

bfredl commented:
Doesn't help. I added a debug message:
```
        self._nvim_mouse_off()
        lib.tickit_term_destroy(tt)
        print("did exit")
```
and `did exit` got printed, so at least the exit code got run. If it is a hint, my $TERM is `rxvt-unicode-256color`. Also it doesn't switch back from the alternative screen (or didn't switch to it in the first place, hard to say).

bfredl commented:
same problem in `xterm` (still emits mouse events and no alt. screen)

leonerd commented:
@bfredl Yah, tickit doesn't shut down the terminal properly in this case it seems.

tarruda commented:
> Also it doesn't switch back from the alternative screen (or didn't switch to it in the first place, hard to say)

@bfredl You are probably missing my libtickit patch from yesterday: https://github.com/neovim/libtickit/commit/0430ba2f43fdf1c31bca66def52a2537c581ade5

Try updating this branch to the latest version and reinstall the dependencies.

(@leonerd, I've been meaning to send you some patches as soon as I rebase our custom branch on top of the lastest window/renderbuffer updates, which I will do later).

bfredl commented:
Wierd, becasue I thought it recompiled tickit earlier today... But anyway, after rebuilding everything again it works :+1:

fwalch commented:
Trying to suspend pynvim with CTRL-Z seems to kill it.

Tranquility commented:
If I start pynvim without a file and open a reasonable large file afterwards the ui gets messed up by moving down.

tarruda commented:
> Trying to suspend pynvim with CTRL-Z seems to kill it.

For now if you start without flags, it will run an embedded instance which gets killed when the parent python process dies(which is what happens when the parent sends a detach command). To detach/attach you need to start pynvim with the -l/-c flags, for example: `pynvim -l /tmp/socket`

> If I start pynvim without a file and open a reasonable large file afterwards the ui gets messed up by moving down.

Is that with the gtk or the tickit UI? Could you send me an example file that would reproduce the bug?

Tranquility commented:
It happens with the tickit UI and you can use the CMakeLists.txt file in the neovim root dir.

tarruda commented:
> It doesn't work with my "csapprox" color scheme (but this is partly my own fault, as my color scheme lanches a python script that manipulates the color palette of urxvt, which somehow causes the frontend to crash) 

Eventually I'd like to handle this in the terminal UI: Try to use RGB colors if the terminal supports it, and if not use color indexes passed to cterm{fg,bg}.

> It happens with the tickit UI and you can use the CMakeLists.txt file in the neovim root dir.

I just did but I'm having trouble spotting this glitch. Could you give me step-by-step instructions, possibly with a vimrc that reproduces the bug?

xueyayang commented:
@tarruda  
Ubuntu 14.04. I want to try the UI and do some test,   Did  i miss something to run the UI?
>sudo apt-get install python-gi
sudo pip install git+https://github.com/tarruda/python-client.git@ui-prototypes

then 
> pynvim

given:
>xue@xue:~$ PKG_CONFIG_PATH=~/xue_work/neovim/.deps/usr/lib/pkgconfig/ pynvim
Traceback (most recent call last):
  File "/usr/local/bin/pynvim", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2749, in <module>
    working_set = WorkingSet._build_master()
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 444, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 725, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 628, in resolve
    raise DistributionNotFound(req)
pkg_resources.DistributionNotFound: futures

Tranquility commented:
I am not able to reproduce it constently. Sometimes it happens sometimes not. But depending on the amount of plugins I remove from my vimrc I see it less often.

tarruda commented:
> Ubuntu 14.04. I want to try the UI and do some test, Did i miss something to run the UI?

I didn't test this on ubuntu 14.04 yet. Can you check if `from gi.repository import GLib, GObject, Gdk, Gtk, Pango, PangoCairo` raises any import errors in the python REPL?

BTW, the PKG_CONFIG_PATH is only required for the terminal UI because a libtickit extension is compiled the first time. If you want to run the gtk version you need to pass the `-g` flag

> I am not able to reproduce it constently. Sometimes it happens sometimes not. But depending on the amount of plugins I remove from my vimrc I see it less often.

Ok I may have seen a redraw glitch, I will keep trying to investigate the problem. Thanks for reporting

Tranquility commented:
And is the background color supposed to work? On my machine it keeps the terminal background color.

tarruda commented:
> And is the background color supposed to work? On my machine it keeps the terminal background color.

It should work. Can you link your color scheme?

tarruda committed:
channel: Make pending_requests a field of the Channel type

This is required to send redraw notifications while a msgpack-rpc call is being
performed by another channel.

tarruda commented:
@Tranquility do you still experience the messed screen after @a107ece?

tarruda committed:
ui: Fix out_flush/ui_write behavior to always flush for abstract_ui

xueyayang commented:
@tarruda 
It seems imports fine:
>xue@xue:~$ python
Python 2.7.6 (default, Mar 22 2014, 22:59:38) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
\>>> from gi.repository import GLib, GObject, Gdk, Gtk, Pango, PangoCairo
\>>> 

And the result of  
>xue@xue:~$ which nvim
/usr/local/bin/nvim

Something else i need to do?

oni-link commented:
Some errors for `:Ag a` on nvim root dir in terminal nvim
```
Traceback (most recent call last):
  File "/home/oni-link/pythonmodules/t/neovim/ui/ui_bridge.py", line 80, in apply_updates
    handler(*args)
  File "/home/oni-link/pythonmodules/t/neovim/ui/tickit_ui.py", line 360, in _nvim_put
    self._screen.put(text, self._attrs)
  File "/home/oni-link/pythonmodules/t/neovim/ui/screen.py", line 93, in put
    cell = self._cells[self.row][self.col]
IndexError: list index out of range
```
and gui nvim
```
Traceback (most recent call last):
  File "/home/oni-link/pythonmodules/t/neovim/ui/ui_bridge.py", line 80, in apply_updates
    handler(*args)
  File "/home/oni-link/pythonmodules/t/neovim/ui/gtk_ui.py", line 227, in _nvim_put
    self._redraw_glitch_fix()
  File "/home/oni-link/pythonmodules/t/neovim/ui/gtk_ui.py", line 487, in _redraw_glitch_fix
    text, attrs = self._screen.get_cursor()
  File "/home/oni-link/pythonmodules/t/neovim/ui/screen.py", line 103, in get_cursor
    return self.get_cell(self.row, self.col)
  File "/home/oni-link/pythonmodules/t/neovim/ui/screen.py", line 99, in get_cell
    return self._cells[row][col].get()
IndexError: list index out of range
```

justinmk commented on src/nvim/term.c:
```diff
@@ -1819,12 +1820,9 @@ static int out_pos = 0;                 /* number of chars in out_buf */
 void out_flush(void)
 {
-  if (out_pos != 0) {
-    /* set out_pos to 0 before ui_write, to avoid recursiveness */
```
is this comment still relevant?

tarruda commented:
@oni-link they are essentially the same error, nvim is sending invalid cursor positions to the UI. I will investigate the cause, thanks.

tarruda committed:
ui: Use ui_linefeed to handle line breaks correctly

ui_linefeed will scroll the screen when it becomes full. This can happen when
executing external commands.

tarruda commented:
@oni-link the `IndexError` should be fixed now

tarruda commented:
> Something else i need to do?

@xueyayang I'm not sure but the error your reported seems could be caused by a wrong distutils installation. A possible solution is to use [pyenv](https://github.com/yyuu/pyenv) to install python and [this script](https://gist.github.com/tarruda/b67472898a7e4bd81572) to install pygtk3

oni-link commented:
The errors for the terminal nvim version are gone, but I still see the old error for the gui version
with `:Ag post src/nvim` and a new ones like

```bash
Traceback (most recent call last):
  File "/home/oni-link/pythonmodules/t/neovim/ui/ui_bridge.py", line 80, in apply_updates
    handler(*args)
  File "/home/oni-link/pythonmodules/t/neovim/ui/gtk_ui.py", line 229, in _nvim_put
    self._screen.put(self._get_pango_text(text), self._attrs)
  File "/home/oni-link/pythonmodules/t/neovim/ui/gtk_ui.py", line 441, in _get_pango_text
    rv = GLib.markup_escape_text(text or '')
  File "/usr/lib64/python2.7/site-packages/gi/overrides/GLib.py", line 452, in markup_escape_text
    return GLib.markup_escape_text(text.decode('UTF-8'), length)
  File "/usr/lib64/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xe5 in position 0: unexpected end of data
```

oni-link commented on src/nvim/ui.c:
```diff
@@ -165,13 +164,39 @@ void ui_cursor_shape(void)
+  for (size_t i = 0; i < ui_count; i++) {
+    UI *ui = uis[i];
+    ui->resize(ui, width, height);
+  }
```
Why not keep the shorter `UI_CALL` instead of the loop?

tarruda replied:
:+1:

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -200,20 +193,21 @@ bool channel_send_event(uint64_t id, char *name, Array args)
     }
+  }  else {
+    // TODO(tarruda): Implement event broadcasting in vimscript
+    broadcast_event(name, args);
```
Are broadcasted events not delayed for a channel with pending requests?

tarruda replied:
:+1:

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -261,7 +255,7 @@ Object channel_send_call(uint64_t id,
   }
 
-  if (!pending_requests) {
+  if (!channel->pending_requests) {
```
`channel` can be freed at this point.

tarruda replied:
:+1:

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -261,7 +255,7 @@ Object channel_send_call(uint64_t id,
 
-  if (!pending_requests) {
+  if (!channel->pending_requests) {
     send_delayed_notifications();
```
This function sends notifications for **all** channels, but does not check `pending_requests == 0` for every channel that has a delayed notification.

tarruda replied:
:+1:

tarruda committed:
shell: When executing command, use screen functions to display output

By calling ui_write directly, the internal screen isn't updated and invalid
bytes aren't handled, which breaks the abstract UI model.

tarruda committed:
ui: Fix redraw bug caused by race conditions with remote clients

Before sending a resize command to the UIs, flush the current output buffer to
ensure no redraw commands for a screen with invalid size are processed.

tarruda commented:
@Tranquility can you test the latest version and report which redraw bugs you still experience?

carlosdcastillo commented:
@tarruda While implementing my own client I started seeing in the client things like the following, as if it was  part of the text I was editing:

, ,1, ,highlight_set,[object Object],put <,!,D,O,C,T,Y,P,E,cursor_goto,0,14,put,H,T,M,L,cursor_goto,0,19,put

I think (I'll try to isolate a minimal example) that:

If you call vim_command redraw! and immediately after you call ui_attach or if in handling data from ui_attach you call vim_command redraw! you leave the data stream in such a way that msgpack interprets some message as put of other things in the stream, dumping the stream in the text area.

tarruda commented:
@carlosdcastillo I just fixed a bug that could result in the client receiving redraw commands for a screen with invalid size, can you test again?

 BTW is your code public? I'd love to see a javascript UI :smile:

tarruda committed:
syntax: Refresh UI when the color scheme changes

fwalch commented:
@tarruda FWIW the redraw error in https://github.com/neovim/python-client/pull/74#issuecomment-69493292 is gone :-)

justinmk commented:
OMG, I'm sorry. That was an accident.

tarruda commented:
> OMG, I'm sorry. That was an accident, GitHub mobile somehow registered 2 clicks in the wrong spot. And there is no way to re-open this PR. Sorry again.

Np, I will just open another PR. BTW, which github mobile client are you using? Have you seen [this one](https://play.google.com/store/apps/details?id=com.gh4a)?

justinmk commented:
> Have you seen this one?

I'll try it, thanks.

tarruda commented:
> @tarruda FWIW the redraw error in neovim/python-client#74 (comment) is gone :-)

Yes, it was fixed by @a16cd73, but I will squash that into another commit. Are you still having [this issue](https://github.com/neovim/python-client/pull/74#issuecomment-69493292)?

fwalch commented:
> Yes, it was fixed by @a16cd73, but I will squash that into another commit. Are you still having this issue?

No; that was actually the one I linked to in my comment.

tarruda commented:
> No; that was actually the one I linked to in my comment.

Ok, that one was fixed by a change in the python client

carlosdcastillo commented:
@tarruda I just pushed my code here: https://github.com/carlosdcastillo/vim-mode

It is still finicky, but I think it should be possible to run. In case it doesn't, here is a video that shows where I am connecting to neovim using abstract-ui-fixes branch: https://www.youtube.com/watch?v=yluIxQRjUCk

tarruda commented:
@carlosdcastillo cool, I will merge the new branch(#1810) today after some extra tests are added. With time there will be more information via the UI protocol which will help embedders that need to have more control over the UI.
